### PR TITLE
[WIP] Porting Linear algebra from CP2K

### PR DESCRIPTION
**Pull request by dithillobothrium**

### CODE REVIEW

**Comment by dithillobothrium** [comment]
I forgot to mention. There is also a change in dbcsr_io. I added possibility to print complex matrix. Before it printed only the real part.

**Comment by dev-zero** [comment]
@dithillobothrium as just discussed: there is still a lot of commented out code in this PR. Please remove them all. If you think some of them are needed (in examples maybe for illustration), please add a comment on top of them.

**Comment by alazzaro** [comment]
retest this please

**Comment by alazzaro** [comment]
retest this please

**Comment by dithillobothrium** [comment]
@dev-zero , can we merge it or retest if coverage tests are changed?

**Comment by alazzaro** [comment]
We cannot merge yet, I have to check it.
Concerning the coverage, arnoldi and most of the dbcsr_vector functions are not tested, we will open an issue for that later... Current deadline for merging is end of the month.

File: examples/tight_binding/tight_binding.F [hunk 1]
```diff
@@ -0,0 +1,1229 @@
...
+    rho = 0.0
+
+    do jj=1,nr
+      add_block = .false. 
+      values(:) = 0.0
+
+      do irft = 1, nrft
+        r2 = rr(:,jj) + rft(:,irft)
+        dvec(:)=r2(:)-r1(:)
+        t1=sqrt(dvec(1)**2+dvec(2)**2+dvec(3)**2)
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: `norm2` again

File: examples/tight_binding/tight_binding.F [hunk 2]
```diff
@@ -0,0 +1,1229 @@
...
+      
+        if (t1 < distance_cutoff .and. ii /= jj) then
+           rho = rho + rhoPart(lambda, t1) 
+           vec_norm(:) = dvec(:) / t1
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: move inside the `if (node_holds ...)` ?

File: examples/tight_binding/tight_binding.F [hunk 3]
```diff
@@ -0,0 +1,1229 @@
...
+   
+  do ii=1,nr
+    r1=rr(:,ii)
+    rho = 0
+    i = (ii - 1) * block_size + 1
+    
+    do jj=1,nr
+      r2=rr(:,jj)
+      dvec(:)=r2(:)-r1(:)
+      t1=sqrt(dvec(1)**2+dvec(2)**2+dvec(3)**2)
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: `norm2`

File: examples/tight_binding/tight_binding.F [hunk 4]
```diff
@@ -0,0 +1,1229 @@
...
+      
+            
+      ! calc onsite
+      ! Es
+      ham(i,i) = diagH(rho, energiesS)
+      diagVal = diagH(rho, energiesP)
+
+      ! Ep
+      do ib= 1,3
+         ham(i + ib, i + ib ) = diagVal
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: maybe `ham(i+1:i+3,i+1:i+3) = diagVal` instead?

File: examples/tight_binding/tight_binding.F [hunk 5]
```diff
@@ -0,0 +1,1229 @@
...
+  integer, intent(in) :: nelec, matrix_size, first_i, first_j, desc_ham(9)
+  real(8), allocatable, intent(inout) :: ham(:,:)
+
+  real(8), allocatable :: work(:)
+  real(8) :: testwork(1)
+  real(8), allocatable :: eigval(:)
+  integer lwork, i
+  integer info,n,ld, desczdef(9)
+  real(8) energy , zdef(2)
+  !real(8), allocatable :: Cmatr(:,:)
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: ?

File: examples/tight_binding/tight_binding.F [hunk 6]
```diff
@@ -0,0 +1,1229 @@
...
+    write(*,'("Error: diagonalisation failed")')
+    write(*,'(" ZHEEV returned INFO = ",I8)') info
+    write(*,*)
+    stop
+  end if
+  
+  if (mynode == 0) then
+    energy=0
+    do i=1,nelec
+      energy = energy + eigval(i)
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: maybe `energy = sum(eigval(1:nelec))`

File: examples/tight_binding/tight_binding.F [hunk 7]
```diff
@@ -0,0 +1,1229 @@
...
+    ! diagonalize
+    call pdsyev('N','U', n, ham, first_i, first_j, desc_ham, eigval, zdef, 0, 0, desczdef, work, lwork, info)
+
+    
+  if (info.ne.0) then
+    write(*,*)
+    write(*,'("Error: diagonalisation failed")')
+    write(*,'(" ZHEEV returned INFO = ",I8)') info
+    write(*,*)
+    stop
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: `error stop`

File: examples/tight_binding/tight_binding.F [hunk 8]
```diff
@@ -0,0 +1,1229 @@
...
+  real(8), allocatable :: eigval(:)
+  integer lwork, i
+  integer info,n,ld
+  real(8) energy
+  !real(8), allocatable :: Cmatr(:,:)
+    
+    info=0
+    n = matrix_size
+    ld = matrix_size
+    lwork=3*n
```
Comment (dev-zero) [inline examples/tight_binding/tight_binding.F]: maybe add a comment?

File: src/la/arnoldi/arnoldi_data_manipulation.f90 [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+#:include 'arnoldi.fypp'
+#:for nametype1, type_prec, real_zero, nametype_zero, type_nametype1, vartype in inst_params_1
+! **************************************************************************************************
+!> \brief ...
+!> \param arnoldi_data ...
+!> \param matrix ...
+!> \param max_iter ...
+! **************************************************************************************************
+  SUBROUTINE setup_arnoldi_data_${nametype1}$ (arnoldi_data, matrix, max_iter)
+     TYPE(arnoldi_data_type)                 :: arnoldi_data
```
Comment (dev-zero) [inline src/la/arnoldi/arnoldi_data_manipulation.f90]: `intent` missing

File: tests/dbcsr_unittest1.F [hunk 10]
```diff
@@ -320,6 +333,8 @@ PROGRAM dbcsr_unittest
    ! clean mp environment
    CALL dbcsr_mp_release(mp_env)
 
+   call dbcsr_mp_grid_remove(mp_env)
```
Comment (dev-zero) [inline tests/dbcsr_unittest1.F]: why is this needed? `mp_release` seems to `grid_remove` internally? Furthermore, the `refcount` of `mp_env` should reach 0 inside the `mp_release` (since we're at the end), meaning that `mp_env%mp` should already be deallocated

File: tests/dbcsr_unittest1.F [hunk 11]
```diff
@@ -90,6 +96,13 @@ PROGRAM dbcsr_unittest
 
    ! run tests
 
+   ! mat x vec ------------------------------------
+   !call dbcsr_mat_vec_multiplies("mat x vec square", group, mp_env, npdims, io_unit, 20, 2, 0.5_dp, &
```
Comment (dev-zero) [inline tests/dbcsr_unittest1.F]: ?

File: tests/dbcsr_test_mat_vec_mult.F [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+      n = 1
+      bs(1) = 1
+      m = matrix_size
+      k = m
+      bs(2) = bs_m
+
+      !open(unit=123,file="mnkb.dat",status="old",form="formatted")
+      !read(123,*) m
+      !read(123,*) k
+      !read(123,*) bs(2)
```
Comment (dev-zero) [inline tests/dbcsr_test_mat_vec_mult.F]: what's with all this commented out code?

File: src/ops/dbcsr_io.F [hunk 13]
```diff
@@ -580,34 +580,63 @@ SUBROUTINE printmat_s(matrix, rows, cols, iunit, title, tr)
...
-            CALL printmat_s(REAL(matrix, KIND=sp), rows, cols, iunit, tr=tr)
+      ENDIF
+      DO r = LBOUND(matrix, 1), UBOUND(matrix, 1)
+         bit_bucket = matrix(r)
+      ENDDO
+      bit_bucket = 0.0_dp
+      DO r = LBOUND(matrix, 1), UBOUND(matrix, 1)
+         bit_bucket = bit_bucket + matrix(r)
+      ENDDO
+      IF (m .GT. 10000) m = 0
```
Comment (dev-zero) [inline src/ops/dbcsr_io.F]: ?

File: tests/dbcsr_test_mat_vec_mult.F [hunk 14]
```diff
@@ -0,0 +1,365 @@
...
+                          sizes_m)
+      CALL dbcsr_dist_bin(col_dist, SIZE(sizes_k), npdims(2), &
+                          sizes_k)
+
+      CALL dbcsr_distribution_new(dist, mp_env, row_dist, col_dist)
+
+      CALL dbcsr_reset_randmat_seed()
+
+!    print *, m,k,bs,sparsity
+!    print *, sizes_m
```
Comment (dev-zero) [inline tests/dbcsr_test_mat_vec_mult.F]: ?

File: src/la/arnoldi/dbcsr_arnoldi_api.F [hunk 15]
```diff
@@ -0,0 +1,545 @@
...
+      INTEGER                                            :: handle, max_iter_internal, nrestarts
+      TYPE(arnoldi_data_type)                            :: my_arnoldi
+      TYPE(dbcsr_p_type), DIMENSION(:), POINTER          :: arnoldi_matrices
+
+      CALL timeset(routineN, handle)
+
+      ! we go in chunks of max_iter_internal, and restart ater each of those.
+      ! at low threshold smaller values of max_iter_internal make sense
+      IF (.TRUE.) max_iter_internal = 16
+      IF (threshold <= 1.0E-3_dp) max_iter_internal = 32
```
Comment (dev-zero) [inline src/la/arnoldi/dbcsr_arnoldi_api.F]: magic values?

File: src/la/arnoldi/dbcsr_arnoldi_api.F [hunk 16]
```diff
@@ -0,0 +1,545 @@
...
+         CALL dbcsr_set(xvec, 0.0_dp)
+         ! Conjugate gradients
+         rsold = vec_dot_vec(rvec, rvec, mpgrp)
+         DO iter = 1, control%max_iter
+            CALL dbcsr_copy(vectors%input_vec, pvec)
+            CALL dbcsr_matrix_colvec_multiply(amat, vectors%input_vec, vectors%result_vec, 1.0_dp, &
+                                              0.0_dp, vectors%rep_row_vec, vectors%rep_col_vec)
+            CALL dbcsr_copy(apvec, vectors%result_vec)
+            pap = vec_dot_vec(pvec, apvec, mpgrp)
+            IF (ABS(pap) < 1.e-24_dp) THEN
```
Comment (dev-zero) [inline src/la/arnoldi/dbcsr_arnoldi_api.F]: magic constant?

File: src/la/arnoldi/dbcsr_arnoldi_api.F [hunk 17]
```diff
@@ -0,0 +1,545 @@
...
+   USE dbcsr_api,                       ONLY: &
+      dbcsr_add, dbcsr_copy, dbcsr_get_block_p, dbcsr_get_info, dbcsr_iterator_blocks_left, &
+      dbcsr_iterator_next_block, dbcsr_iterator_start, dbcsr_iterator_stop, dbcsr_iterator_type, &
+      dbcsr_p_type, dbcsr_release, dbcsr_set, dbcsr_type, dbcsr_create_col_vec_from_matrix,&
+      dbcsr_create_replicated_col_vec_from_matrix, dbcsr_create_replicated_row_vec_from_matrix,&
+      dbcsr_matrix_colvec_multiply
+   !USE dbcsr_vector,                    ONLY: create_col_vec_from_matrix,&
+   !create_replicated_col_vec_from_matrix,&
+   !create_replicated_row_vec_from_matrix,&
+   !dbcsr_matrix_colvec_multiply
```
Comment (dev-zero) [inline src/la/arnoldi/dbcsr_arnoldi_api.F]: ?

File: src/ops/dbcsr_io.F (hunk 13 above)
Comment (dithillobothrium) [reply src/ops/dbcsr_io.F]: what is wrong here?

File: examples/tight_binding/tight_binding.F [hunk 18]
```diff
@@ -0,0 +1,1229 @@
...
+   read (2, *) nr
+   allocate(rr(3,nr))
+   do j=1,nr
+       read (2, *) rr(:,j) 
+   enddo
+   close(2)
+
+   rr = rr / 0.529
+   
+   sys_dim = 3 ! fixed!
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: 3 dimensions of the system. Because before it was changeable, now not.

File: examples/tight_binding/tight_binding.F [hunk 19]
```diff
@@ -0,0 +1,1229 @@
...
+          s_dense(nr*block_size, nr*block_size))
+      
+      call build_ham_blocked_dense(ham_dense, rr, nr, block_size, dist_cutoff, rft, nrft)
+      !call build_ham_blocked_dense_no_trans(ham_dense2, rr, nr, block_size, dist_cutoff)
+      !call build_S_blocked_dense(s_dense, rr, nr, block_size, dist_cutoff)
+      
+      if (nr <= 10) then
+         write(exFmt,'("(",I0,"(F8.4))")') nr*block_size
+         do i=1,nr*block_size
+         write(*,exFmt) (ham_dense(i,j), j=1,nr*block_size)
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: it prints a part of matrix allocating on each node. It happens only if small system is used. And it is convenient to check matrix opening xterm window for each MPI rank.

File: examples/tight_binding/tight_binding.F [hunk 20]
```diff
@@ -0,0 +1,1229 @@
...
+   if (nr < 500) then
+      
+       if (mynode == 0)  print *,"Diagonalize dense hamiltonian for reference"
+
+      allocate(ham_dense(nr*block_size, nr*block_size),&
+          s_dense(nr*block_size, nr*block_size))
+      
+      call build_ham_blocked_dense(ham_dense, rr, nr, block_size, dist_cutoff, rft, nrft)
+      !call build_ham_blocked_dense_no_trans(ham_dense2, rr, nr, block_size, dist_cutoff)
+      !call build_S_blocked_dense(s_dense, rr, nr, block_size, dist_cutoff)
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: It is an example. One can build another Hamiltonian matrix and overlapping matrix. I need to add comments for that.

File: examples/tight_binding/dbcsr_scalapack.F [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+    il  = (im1/(np*nb))*nb + mod(im1,nb) + 1
+    return
+  end subroutine g2l
+  !
+  ! convert local index to global index in block-cyclic distribution
+  !
+  subroutine l2g(il, p, n, np, nb, i)
+    integer :: il   ! local array index, input
+    integer :: p    ! processor array index, input
+    integer :: n    ! global array dimension, input
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/dbcsr_scalapack.F]: what?

File: src/ops/dbcsr_io.F (hunk 13 above)
Comment (dev-zero) [reply src/ops/dbcsr_io.F]: magic constants

File: examples/tight_binding/dbcsr_scalapack.F (hunk 21 above)
[the remark this answers is not in the document]
Comment (dev-zero) [reply examples/tight_binding/dbcsr_scalapack.F]: `n`n is unused...

File: src/ops/dbcsr_io.F (hunk 13 above)
Comment (dithillobothrium) [reply src/ops/dbcsr_io.F]: I copied it from the above `print_matrix_s` , it is the same there.

File: examples/tight_binding/dbcsr_scalapack.F [hunk 22]
```diff
@@ -0,0 +1,378 @@
...
+  ! convert global index to local index in block-cyclic distribution
+  !
+  subroutine g2l(i, n, np, nb, p, il)
+    integer :: i    ! global array index, input
+    integer :: n    ! global array dimension, input
+    integer :: np   ! processor array dimension, input
+    integer :: nb   ! block size, input
+    integer :: p    ! processor array index, output
+    integer :: il   ! local array index, output
+    integer :: im1
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/dbcsr_scalapack.F]: I did not understand. Should I add intent in out? It seems it is working like that.

File: examples/tight_binding/tight_binding.F [hunk 23]
```diff
@@ -0,0 +1,1229 @@
...
+
+   CALL mpi_comm_rank(MPI_COMM_WORLD, mynode, ierr)
+   IF (ierr /= 0) STOP "Error in MPI_Comm_rank"
+
+   if (mynode == 0) WRITE (*, *) ' numnodes ', numnodes
+
+  prow = npdims(1)
+  pcol = npdims(2)
+
+   ALLOCATE(pgrid(0:prow-1, 0:pcol-1))
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: it is a requirement

File: examples/tight_binding/tight_binding.F (hunk 2 above)
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: no, it will not work, I think

File: examples/tight_binding/tight_binding.F (hunk 4 above)
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: will it give me a matrix 3x3, not diagonal?

File: examples/tight_binding/tight_binding.F (hunk 8 above)
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: only like "see lapack manual"

File: src/la/arnoldi/arnoldi_data_manipulation.f90 (hunk 9 above)
Comment (dithillobothrium) [reply src/la/arnoldi/arnoldi_data_manipulation.f90]: Why do we need intents if it works without them?

File: tests/dbcsr_unittest1.F (hunk 10 above)
Comment (dithillobothrium) [reply tests/dbcsr_unittest1.F]: No, we discussed it with Alfio already. It should be here otherwise it reports about comm leaks. Also, it cannot be built in dbcsr_mp_release, because some of the tests fail, probably because of some manipulations in Cannon's files. Maybe it should be put before `dbcsr_mp_release`. But I dont remember.

File: src/la/arnoldi/dbcsr_arnoldi_api.F (hunk 15 above)
Comment (dithillobothrium) [reply src/la/arnoldi/dbcsr_arnoldi_api.F]: there are several empirical values. 16 32 64 1.0E-3_dp  1.0E-4_dp. I think it does not make too much sense to create for all of them a constant variable. The function is small itself and everything is clear.

File: examples/tight_binding/tight_binding.F [hunk 24]
```diff
@@ -0,0 +1,1229 @@
...
+
+  if (mynode==0) print *,"copy dbcsr to blacs matrix"
+
+  call copy_dbcsr_to_blacs(matrix_ks, myA, nb, nb, n, n)
+  
+
+  if (nr < 10) then
+         write(exFmt,'("(",I0,"(F8.4))")') myrows
+         do i=1,mycols
+         write(*,exFmt) (myA(i,j), j=1,myrows)
```
[the remark this answers is not in the document]
Comment (dithillobothrium) [reply examples/tight_binding/tight_binding.F]: removed

**Comment by dithillobothrium** [comment]
Why did it not find the mpi module?

/home/travis/build/cp2k/dbcsr/tests/dbcsr_la_test.F:40:7:
    
use mpi

**Comment by alazzaro** [comment]
You have to use our wrappers in tests since the library has to work with and without MPI

**Comment by dithillobothrium** [comment]
> You have to use our wrappers in tests since the library has to work with and without MPI

Ok, I will change it
